### PR TITLE
Streamline performance testing setup

### PR DESCRIPTION
**Pull request by roypat**
## Changes

This PR streamlines our existing performance testing setup by
- Removing nonsensical values (e.g. pkt_loss for network latency test, which is always 0 in every baseline, and has never been non-zero)
- Removing highly correlated tests (e.g. block tests with mode read and randread had a correlation of r = 0.97 across all runs since February)
- For throughput tests, remove suboptimal test setups.

It also changes the way test parameterization is handled, by moving configuration from the baseline files to `@pytest.mark.parameterize` decorators. This will make it more obvious how far a running performance test run has progressed, as well as make it easier to see what exact configuration failed at a glance.

No changes in this PR require regathering of baseline - any changes that would involve new baselines will be made in follow ups. 

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by roypat** [comment]
Correlation for block device iops/throughput

Screenshot 2023-05-23 at 12-03-41 performance … - JupyterLab

Correlation for different fio modes: 

Screenshot 2023-05-24 at 09-12-13 performance … (2) - JupyterLab

File: tests/integration_tests/performance/test_vsock_throughput.py [hunk 1]
```diff
@@ -213,66 +226,74 @@ def consume_iperf_output(cons, result):
...
     Test vsock throughput for multiple vm configurations.
 
     @type: performance
     """
 
+    # We run bi-directional tests only on uVM with more than 2 vCPus
+    # because we need to pin one iperf3/direction per vCPU, and since we
+    # have two directions, we need at least two vCPUs.
+    if mode == "bd" and vcpus < 2:
+        pytest.skip("bidrectional test only done with at least 2 vcpus0")
```
Comment (pb8o) [inline tests/integration_tests/performance/test_vsock_throughput.py:295]: nit: there's a 0 at the end.

**Review by pb8o — APPROVED** [review]

**Review by pb8o — APPROVED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [workspace]
-members = ["src/firecracker", "src/jailer", "src/seccompiler", "src/rebase-snap", "src/cpu-template-helper"]
+members = ["src/cpu-template-helper", "src/firecracker", "src/jailer", "src/rebase-snap", "src/seccompiler"]
```
Comment (JonathanWoollett-Light) [inline Cargo.toml]: I think this diff is also due to cargo-sort.

File: tests/integration_tests/performance/test_network_tcp_throughput.py [hunk 3]
```diff
@@ -228,78 +242,75 @@ def consume_iperf_tcp_output(cons, result, vcpus_count):
...
+    prod = producer.LambdaProducer(produce_iperf_output, prod_kwargs)
+    return cons, prod, f"{env_id}/{iperf3_id}"
 
 
 @pytest.mark.nonci
 @pytest.mark.timeout(3600)
 @pytest.mark.parametrize("vcpus", [1, 2])
+@pytest.mark.parametrize(
+    "payload_length", ["DEFAULT", "1024K"], ids=["pDEFAULT", "p1024K"]
+)
```
Comment (JonathanWoollett-Light) [inline tests/integration_tests/performance/test_network_tcp_throughput.py:290]: Given [test: remove p=1024 case from vsock throughput perf test](https://github.com/firecracker-microvm/firecracker/pull/3736/commits/fdce3b8d8c6bbde994aa4bfe7402f06d11df2b4f) shouldn't this not contain `1024k`?

File: src/firecracker/Cargo.toml [hunk 4]
```diff
@@ -18,7 +18,7 @@ api_server = { path = "../api_server" }
 logger = { path = "../logger" }
 mmds = { path = "../mmds" }
 seccompiler = { path = "../seccompiler" }
-snapshot = { path = "../snapshot"}
```
[the remark this answers is not in the document]
Comment (roypat) [reply src/firecracker/Cargo.toml]: Ahh, I think so - I probably either messed up a rebase, or accidentally commited suggestions from my pre-commit hook before cargo-sort actually made it to main. Either way, fixed now

File: tests/integration_tests/performance/test_vsock_throughput.py (hunk 1 above)
Comment (roypat) [reply tests/integration_tests/performance/test_vsock_throughput.py]: huh, I thought I'd removed that, fixed!

File: Cargo.toml (hunk 2 above)
Comment (roypat) [reply Cargo.toml]: Yeah, no idea how this one snuck in :( Should be fixed now though!

File: tests/integration_tests/performance/test_network_tcp_throughput.py (hunk 3 above)
Comment (roypat) [reply tests/integration_tests/performance/test_network_tcp_throughput.py:279]: So we used to test _three_ payload lengths: 1024, 1024**K** and DEFAULT. That commit only removes the first one

**Comment by roypat** [comment]
Note: the latest force-push dropped the commit that moves parameterization of the block device from baseline file to pytest. A brief investigation with @pb8o showed that this change results in all tests running in fresh microvms, whereas currently the same microvm is recycled for all tests. This has impact on the results (with a fresh VM achieving better performance). Therefore, this change is moved to a later PR, as it will require regathering of baselines.

**Review by pb8o — APPROVED** [review]

**Review by JonathanWoollett-Light — COMMENTED** [review]
Nit: In the commit `test: remove read and readwrite fio modes from block perf test` the body contains:

> In a later commit, we'll ...

I think we should try to avoid describing 1 commit by referencing another, ideally they should make sense in isolation.

**Comment by roypat** [comment]
> Nit: In the commit `test: remove read and readwrite fio modes from block perf test` the body contains:
> 
> > In a later commit, we'll ...
> 
> I think we should try to avoid describing 1 commit by referencing another, ideally they should make sense in isolation.

It does make sense in isolation (it does contain the reason for why this specific change was made). The other part is simply added context.

> I would also say the same about commit `test: remove p=1024 case from vsock throughput perf test ` with
> 
> > As in the previous commit, ...

This one I can see, I'll keep it in mind going forward. Will keep it like this though since CI is green and this PR has approvals

**Pull request merged**